### DISCUSSION TITLE
Solver seems to start breaking hard constraint when I add another planning variable

### DISCUSSION THREAD

**Discussion by matthewjd24** [TimefoldAI/timefold-solver#175] [General]
Here is a diagram of my domain model: 
![image](https://github.com/TimefoldAI/timefold-solver/assets/27173297/d4db9673-20c1-4b4b-a362-73bd7902855f)

The issue I have is, when I add in the Mold as a planning variable, the solver starts breaking my hard constraint (the only one I've set up so far). Yet the HardSoftScore result is still `best score (0hard/0soft)`.

The hard constraint applies a penalty when two jobs are assigned to the same Cycle (checked by getTimeslotID). The hard constraint:

```
private Constraint oneJobAtATimeOnMachine(ConstraintFactory constraintFactory) {
        //tried rewriting this constraint in different ways but did not help
        return constraintFactory
            .forEach(Job.class)
            .filter(job -> job.getCycle() != null)
            .join(Job.class,
                Joiners.equal(Job::getTimeslotID),
                Joiners.filtering((job1, job2) -> job1.getId() != job2.getId()))
            .penalize(HardSoftScore.ONE_HARD)
            .asConstraint("Room conflict");
    }
```

I've rewritten this constraint multiple times in the hope that it's not behaving as I expect, but this issue persists. 

My PlanningEntity (Job):
```
@PlanningEntity
public class Job {

    @PlanningId
    private Long id;

    public String jobName;
    public Integer width;
    public Integer height;
    public Integer weight;

    @PlanningVariable(nullable = true)
    private Cycle cycle;
    @PlanningVariable(nullable = true)
    private Mold mold;
```

My PlanningSolution (TimeTable):
```
@PlanningSolution
public class TimeTable {

    @ValueRangeProvider
    @ProblemFactCollectionProperty
    private List<Cycle> timeslotList;
    @ValueRangeProvider
    @ProblemFactCollectionProperty
    private List<Mold> moldList;
    @PlanningEntityCollectionProperty
    private List<Job> lessonList;

    @PlanningScore
    private HardSoftScore score;
```

Things I've tried:

- Run in environment modes FULL_ASSERT and NON_INTRUSIVE_FULL_ASSERT. No exceptions occur during either and the same issue persists.
- Rewrite constraint.
- Verify in multiple ways that multiple jobs are indeed being assigned to the same cycle.
- Remove all constraints except the hard constraint here and my soft constraint that penalizes Cycles with no assigned Jobs.
- Changing `@PlanningVariable(nullable = true)` to `@PlanningVariable(valueRangeProviderRefs = {"moldRange"}, nullable = true)` in the Job class and changing `@ValueRangeProvider` to `@ValueRangeProvider(id = "moldRange")` in my PlanningSolution.

If I remove the Mold as a planning variable, it starts obeying the constraint again. I do not know why. I have published my project in a repository here: https://github.com/matthewjd24/OptaPlanner-Scheduler

**Comment by matthewjd24** [comment]
Making the mold property of the Job class not nullable seems to fix it. I'm not sure why. It's not an accurate representation of my domain but I can work around it.

**Comment by nicksi** [comment]
According to documentation (https://timefold.ai/docs/timefold-solver/latest/constraint-streams/constraint-streams.html#constraintStreamsForEach) the  .forEach(Job.class) tracks only planning entities with non Null planning variable values. Try .forEachIncludingNullVars(Job.class) instead

**Reply by matthewjd24** [reply]
Thank you
